test/spec/modules/33acrossAnalyticsAdapter_spec.js

```diff
@@ -0,0 +1,2368 @@
+        it('it logs an error message', function() {
+          analyticsAdapter.enableAnalytics({
+            provider: '33across',
+            options: {
+              endpoint: faker.internet.url()
+            },
+          });
+
+          assert.calledOnce(log.error);
+          assert.calledWithExactly(log.error, 'No partnerId provided for "options.pid". No analytics will be sent.');
+        });
+      });
+    });
+
+    context('When pid and endpoint are given', function() {
+      context('and an invalid timeout config value is given', function() {
+        it('it logs an info message', function() {
+          [ null, 'foo', -1 ].forEach((value, index) => {
+            analyticsAdapter.enableAnalytics({
+              provider: '33across',
+              options: {
+                pid: 'test-pid',
+                endpoint: 'http://test-endpoint',
+                timeout: value
+              },
+            });
+            analyticsAdapter.disableAnalytics();
+
+            assert.called(log.info);
+            assert.calledWithExactly(log.info.getCall(index), 'Invalid timeout provided for "options.timeout". Using default timeout of 3000ms.');
+          });
+        });
+      });
+
+      it('logs any received GAM slotRenderEnded event', function() {
+        analyticsAdapter.enableAnalytics({
+          provider: '33across',
+          options: {
+            pid: 'test-pid',
+            endpoint: faker.internet.url(),
+          },
+        });
+
+        performAuctionWithMissingBidWon();
+
+        window.googletag.cmd.forEach(fn => fn());
+
+        const { gam } = getMockEvents();
+        const [ gEvent ] = gam.slotRenderEnded;
+
+        mockGpt.emitEvent('slotRenderEnded', gEvent);
+
+        assert.calledWithExactly(log.info, 'slotRenderEnded', `adUnitPath: ${gEvent.slot.getAdUnitPath()}`, `slotElementId: ${gEvent.slot.getSlotElementId()}`);
+      });
+    });
+  });
+
+  describe('Event Handling', function() {
+    beforeEach(function() {
+      this.defaultTimeout = 3000;
+      this.enableAnalytics = (options) => {
+        analyticsAdapter.enableAnalytics({
+          provider: '33across',
+          options: {
+            endpoint: 'http://test-endpoint',
+            pid: 'test-pid',
+            timeout: this.defaultTimeout,
+            ...options
+          },
+        });
+      }
+    });
+
+    context('when an auction is complete', function() {
+      context('and the AnalyticsReport is sent successfully to the given endpoint', function() {
+        it('it logs an info message', function() {
+          const endpoint = faker.internet.url();
+          this.enableAnalytics({ endpoint });
+
+          sandbox.stub(navigator, 'sendBeacon').returns(true);
+
+          performStandardAuction();
+          sandbox.clock.tick(this.defaultTimeout + 1);
+
+          assert.calledOnce(navigator.sendBeacon);
+          assert.calledWith(navigator.sendBeacon, endpoint, sinon.match.string);
```

Comment:
in order to remove the number of assertions, I think it's better to user `withArgs` in the stub

e.g.
```
sandbox.stub(navigator, 'sendBeacon')
  .withArgs(endpoint, sinon.match.string))
  .returns(true);
```

sendBeacon will return true only when its called with those specific specs.

and that way you could remove  this assertion 
```
assert.calledWith(navigator.sendBeacon, endpoint, sinon.match.string);
```

Reply:
also to be more precise, let's use `JSON.stringify(getStandardAnalyticsReport())` instead of sinon.match.string

Reply:
Regarding `.withArgs` vs `calledWith`: 

I like the idea of matching on anything rather than for the specific expected value for the failure case: if the input to navigator.sendBeacon doesn't match the expected value, then the assertion can gather the diff between what we expected and what we got.

Regarding precision of the assertion:

My intention was to write a matcher (effectively a validator for the schema) where we could assert that the report matched on the defined traits of the report, rather than doing an exact match on the input/output, but I just didn't get to it. That would give us a better test of the things we care about (does our report contain the essential pieces, and are they of the correct type and formatting?) without creating a lot of churn when we want to make small changes to the mocks...but maybe we can save that for fuzz testing. I'm comfortable switching this back to using an exact string match.

Reply:
A change in this logic has to break the unit test.. e.g. a developer during experimental testing decides to send an empty string and forgets to revert this change.. the unit test will still succeed if we keep the sinon.match.string

Reply:
I added a report validation function to align with my original intentions here.